### PR TITLE
gsc: a member on an element from .Keys/.Values binds, and a map over a nullable value type emits verifiable IL (#4033, #4035)

### LINKED ISSUE #4033: gsc: a member call on an element from `for v in dict.Values` crashes with GS9998 when the value is a constructed generic
Found while fixing #4023 (measured, not fixed there — it affects BOTH spellings equally, so it is not the map/`Dictionary` inconsistency that issue is about).

Iterating a dictionary's `.Values` and then calling a member on the loop element reports the internal-compiler-error `GS9998` when the value type is a constructed generic.

## Repro

```gs
package Demo
import System
import System.Collections.Generic

func main2() {
    var g = Dictionary[string, List[int32]]()
    g["b"] = List[int32]()
    g["b"].Add(5)
    for lists in g.Values {
        Console.WriteLine(lists.Count.ToString())
    }
}

main2()
```

```
error GS9998: EmitDiagnosticException: Internal compiler error: an unresolved (error)
expression reached code emission. This usually indicates an unsupported or mis-bound
construct that earlier phases failed to report.
        Console.WriteLine(lists.Count.ToString())
```

## Measured matrix (current `main`, `565b9a04`, and on the #4023 branch — identical on both)

| program | result |
| --- | --- |
| `for lists in Dictionary[string, List[int32]]().Values { lists.Count }` | **`GS9998`** |
| `for lists in map[string, List[int32]]{}.Values { lists.Count }` | **`GS9998`** (after #4023; `GS0158` before, because `.Values` did not bind at all) |
| `g["b"].Count` — the same member on the same element, reached through the indexer | green |
| `for v in m.Values { v.Name }` where the value is a non-generic imported class | green |
| `for v in m.Values { v.Count.ToString() }` — counting the collection, not the element | green |

So it is specific to **reaching a member on an element obtained by iterating `.Values`, when the value type is a constructed generic**. The two spellings agree, which is what puts it outside #4023: that issue is about the map spelling losing members the `Dictionary` spelling keeps.

The message says an error node reached emission with nothing having reported it, so there is a missing diagnostic underneath as well.

## Expected

`lists.Count` binds on the loop element, exactly as `g["b"].Count` does on the same value reached through the indexer. Failing that, a real diagnostic rather than `GS9998`.

## Notes

- Not channel business, so no ADR-0174 erratum.
- `.Keys` is unaffected in the measured rows (the key was `string` in all of them); a generic KEY type has not been probed.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01UVZ9GDN9g4NVF47GRNpWkL

### LINKED ISSUE #4035: gsc: a `map[K, V?]` over a nullable value type emits unverifiable IL — the backing Dictionary drops the `Nullable<>`
Found while fixing #4023 (measured, not fixed there — it reproduces on `map[string, int32?]`, a shape that #4023's fix cannot reach, and it is a different mechanism).

A `map[K, V?]` over a nullable VALUE type emits IL that does not verify and crashes at runtime with `InvalidProgramException`. The map's backing `Dictionary<K, V>` is closed over the BARE underlying type while the values on the stack are `Nullable<V>`.

## Repro

```gs
package Demo
import System

func main2() {
    var m = map[string, int32?]{"a": 1}
    var x = m["a"]
    Console.WriteLine(m.Count.ToString())
    Console.WriteLine(x.ToString())
    Console.WriteLine(m.ContainsKey("a").ToString())
}

main2()
```

Compiles clean, then:

```
Unhandled exception. System.InvalidProgramException: Common Language Runtime detected an invalid program.
   at Demo.<Program>.main2()
```

`ilverify`:

```
[IL]: Error [StackUnexpected]: [... : Demo.<Program>::main2()][offset 0x00000011]
      [found value '[S.P.CoreLib]System.Nullable`1<int32>'][expected Int32] Unexpected type on the stack.
[IL]: Error [StackUnexpected]: [... : Demo.<Program>::main2()][offset 0x0000001F]
      [found address of '[S.P.CoreLib]System.Nullable`1<int32>'][expected address of Int32] Unexpected type on the stack.
```

## Mechanism

`NullableTypeSymbol`'s constructor passes its underlying type's `ClrType` straight through:

```csharp
private NullableTypeSymbol(TypeSymbol underlyingType)
    : base(ComputeDisplayName(underlyingType), underlyingType.ClrType)
```

so `int32?.ClrType` is `System.Int32`, not `System.Nullable<System.Int32>`. `MapTypeSymbol.MakeClrType` reads `valueType.ClrType` directly, so `map[string, int32?]` is backed by `Dictionary<string, int>` while the literal, the indexer's `out` slot and the loop variable all carry a `Nullable<int>`. The compiler already has the correction for this — `NullableTypeSymbol.GetEffectiveClrType` / `NullableLifting.GetEffectiveClrType`, which `SequenceTypeSymbol.MakeClrType` and `AsyncSequenceTypeSymbol.MakeClrType` both use — but the map's does not.

## Measured

| program | result |
| --- | --- |
| `map[string, int32?]{"a": 1}` then `m["a"]`, `.Count`, `.ContainsKey` | compiles; **2 ilverify errors**, `InvalidProgramException` at runtime |
| `map[string, DateTime?]{}` then `m["a"] = …`, `m["a"]`, `.Count`, `.ContainsKey` | compiles; **2 ilverify errors** (the same two, over `Nullable<DateTime>`); happens to run. Measured on the #4023 branch only — before it, this shape did not reach emit at all, so the nullable defect was masked by #4023 |
| `map[string, int32]` — the non-nullable sibling | green, ilverify-clean |
| `Dictionary[string, int32?]()` — the `Dictionary` spelling | not yet probed |

`int32?` is the important row, and it is the one measured on BOTH `565b9a04` and the #4023 branch: both its key and its value have host-runtime CLR types, so it never goes near #4023's cross-reflection-context path and behaves identically before and after that fix. `DateTime?` shows the same two errors for the same reason, once #4023 lets it reach emit at all.

## Expected

`map[K, V?]` is backed by `Dictionary<K, Nullable<V>>`, so the values the literal, the indexer and `for k, v in m` put on the stack match the dictionary's own value type, and the emitted IL verifies.

## Notes

- Nullable REFERENCE values (`map[string, string?]`) are not affected — the annotation has no distinct runtime type there.
- Not channel business, so no ADR-0174 erratum.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01UVZ9GDN9g4NVF47GRNpWkL

### PR DESCRIPTION
**Pull request by DavidObando**
Two map/dictionary emit bugs, one per commit. **They do not share a root cause.** Both are a consumer forgetting that `NullableTypeSymbol` is a *wrapper*, but #4033 is a **substitution** slip in property-type derivation (`MemberLookup.GetClrPropertyTypeSymbol`) and #4035 is a **construction** slip in `MapTypeSymbol.MakeClrType`. Measured: each commit's witness revert leaves the other issue's rows unchanged. They are paired only because each was found while fixing #4023.

Branched off `3e5103de`, which is current `origin/main`; the task named `acb89d83`, but the two commits since it touch CI/build only, no `src/`.

Fixes #4033
Fixes #4035

---

## #4033 — a member on an element from `.Keys` / `.Values`

`for v in m.Values` followed by a **property** read on the loop element reported `GS0158`, or the internal-compiler-error `GS9998` when the element was a constructed generic — identically for the `map` and the `Dictionary` spelling, which is what put it outside #4023.

### Root cause

`Dictionary<,>.Values` is declared `[Nullable({1, 0, 0})]` (measured off the reference assemblies). Byte 0 is `ValueCollection` itself; the two zeroes sit at the `TKey`/`TValue` positions — those are type **parameters**, so the byte is a placeholder for whatever is substituted, not an annotation.

`GetClrPropertyTypeSymbol` only projected the receiver's symbolic arguments through the open property when `projectOnlyWhenSymbolicallyRequired` was satisfied by one of two narrow probes, so a **closed** `Dictionary[string, string]` fell back to reading those placeholder bytes as concrete positions. #1354's rule — "an unannotated reference position means `T?`" — then applied to a slot that has no annotation to be oblivious *about*, and `.Keys`/`.Values` surfaced a `string?` element. Because G# deliberately requires narrowing or `?.` before a **property** read on an explicit `T?` (but not before a **method** call), the loop body then lost the element's whole CLR property/field surface.

The machinery to do this right already existed: `NullableFlagsBuilder.MergeDeclarationNullability` carries an explicit, documented carve-out taking a type-parameter position's nullability from the receiver's *argument* rather than from the declaration byte — which is exactly why the **open** `map[K, V].Keys` spelling was always right (#3311 pins it). The gate simply never let a closed receiver reach it. The fix projects whenever the open property type mentions a parameter at all and leaves the merge deciding concrete positions, so **#1354 is untouched**.

The map spelling needed a second half: a closed `map[K, V]` arrives as a `MapTypeSymbol`, which is neither an `ImportedTypeSymbol` nor a `StructSymbol`, so `GetProjectionReceiverImportedType` answered `null` and the map spelling stayed broken after the `Dictionary` spelling was fixed. `TryGetClosedMapProjectionView` surfaces the `Dictionary[K, V]` view ADR-0104 says it *is*, taking its open definition from the map's **own** `ClrType` because #4023 may have built that dictionary inside a `MetadataLoadContext`.

### The issue's diagnosis is wrong on two axes, measured

- It names the failing ingredient as "when the value is a **constructed generic**". The real predicate is "a **property** (not a method) on an element from `.Keys`/`.Values`". `for v in Dictionary[string, string]().Values { v.Length }` fails identically on `3e5103de` with a plain `string` value, and the issue's own "green" row `for v in m.Values { v.Name }` is green only because it was never run. Conversely `x.Add(7)` — a *method* on the very element the issue reports as broken — was already green, because a call never consults the receiver gate.
- It under-reports the blast radius: `.Keys` is recorded as "unaffected" and a generic key as "not probed". Measured, **`.Keys` is affected identically**, a **generic key** is affected, and so is **`SortedDictionary[K, V].Values`** — every nested `KeyCollection`/`ValueCollection`. Non-nested generic collections (`List`, `HashSet`, `Queue`, `Stack`, `LinkedList`) were never affected and are control rows.

Both corrections are posted on the issue.

### The old reading was not faithful either

Measured on `3e5103de`: a `Dictionary<string, string?>` and a `Dictionary<string, string>` returned from the **same** nullable-enabled C# assembly both surfaced a `string?` element from `.Values`, because both read the same placeholder — while both dictionaries' own indexer and `for k, v` already surfaced `string`. So the fix does not lose information the compiler had; it makes `.Values` agree with the two paths that were already right.

### Witness (ADR-0154), run as the suite

| row | on `3e5103de` |
| --- | --- |
| `for lists in Dictionary[string, List[int32]]().Values { lists.Count }` | GS9998 |
| `for lists in map[string, List[int32]]{}.Values { lists.Count }` | GS9998 |
| `for v in Dictionary[string, string]().Values { v.Length }` | GS0158 |
| `for k in map[string, string]{}.Keys { k.Length }` | GS0158 |
| `for k in Dictionary[string, int32]().Keys { k.Length }` | GS0158 |
| `for k in Dictionary[List[int32], string]().Keys { k.Count }` | GS9998 |
| `for v in SortedDictionary[string, string]().Values { v.Length }` | GS0158 |
| `var vs = g.Values; for x in vs { x.Count }` | GS9998 |
| `for b in Factory.BoxValues().Values { b.Label }` | GS0158 |
| `Factory.NullableValues()` across indexer, `for k, v` and `.Values` | GS0158 (on the `.Values` path) |
| `for v in Factory.NonNullValues().Values { v.Length }` | GS0158 |
| the 5 control rows and the 3 rejected rows | green |

Reverting `src/Core/CodeAnalysis/Binding/MemberLookup.cs` to `3e5103de`, rebuilding, and running `--filter FullyQualifiedName~Issue4033`:

```
Failed! - Failed: 11, Passed: 8, Skipped: 0, Total: 19
```

**11 red / 8 green before, all 19 green after.** Every accepted row turns red on the revert and every control and rejected row stays green, so the fix is what moved them.

---

## #4035 — a `map[K, V?]` over a nullable value type

Compiled clean, then ilverify rejected the IL with `StackUnexpected` and the runtime threw `InvalidProgramException`: the backing `Dictionary` was closed over the **bare** underlying type while every value on the stack was a `Nullable<V>`.

### Root cause

`NullableTypeSymbol` passes its underlying type's `ClrType` straight through — `int32?.ClrType` is `System.Int32` — because for a nullable *reference* type the annotation has no distinct runtime type. `MapTypeSymbol.MakeClrType` read `ClrType` directly, so `map[string, int32?]` was backed by `Dictionary<string, int>`. The compiler already owns the correction — `NullableLifting.GetEffectiveClrType`, which `SequenceTypeSymbol.MakeClrType` and `AsyncSequenceTypeSymbol.MakeClrType` both call — and the map's construction simply never called it.

### Two arms, not one

`GetEffectiveClrType` builds the wrapper with the **host** `typeof(Nullable<>)`. For `map[string, DateTime?]`, whose underlying `DateTime` is a `MetadataLoadContext` type, that is exactly the `TypeBuilderInstantiation` stand-in #4023 is about — not null, so every `ClrType != null` gate takes it for a reflectable dictionary, and its own assembly cannot name its context. The reference-context arm therefore projects the open `Nullable<>` alongside the open `Dictionary<,>` and closes it over the already-projected underlying, inside the one context. The `DateTime?` and `ImportedPoint?` rows are what separate the complete fix from the naive `GetEffectiveClrType` swap.

### A third site, found by measurement

With the construction fixed, `for v in m.Values` over a `map[string, int32?]` **still** failed ilverify: `MemberLookup.ProjectSymbolicArgToErasedClr` erased the symbolic `int32?` argument through the same bare `ClrType`, closing `ValueCollection` over `int32` while the receiver was a `Dictionary<string, Nullable<int32>>`. Same dropped wrapper, one erasure further out; `BuildErasedNullableInContext` is its `Nullable<>` analogue of the existing `BuildErasedTupleInContext`.

`MethodBodyEmitter`'s `TryGetValue` fallback reads the effective type too. That one is a **correct-by-inspection consistency change, not a measured fix**: it fires only when the carrier lacks two generic arguments, and no row in the suite reaches it. It is kept in step so the two cannot describe the same map differently.

### The issue leaves one row unprobed and it matters

It records the `Dictionary[string, int32?]` spelling as "not yet probed". Measured on `3e5103de`, that spelling is **green** — so the two spellings of one type (ADR-0104) genuinely disagreed, and `var d Dictionary[string, int32?] = map[string, int32?]{"a": 2}` failed to verify with a found `Dictionary<string,int32>` where a `Dictionary<string,Nullable<int32>>` was expected. That row is in the suite; it proves the fix made the spellings **agree** rather than merely moving the map.

### Witness (ADR-0154), run as the suite

| row | on `3e5103de` |
| --- | --- |
| `map[string, int32?]{"a": 1}` then `m["a"]`, `.Count`, `.ContainsKey` | 2 ILVerify `StackUnexpected`; `InvalidProgramException` |
| `m["a"] = 7` then read it back | 2 `StackUnexpected`; `InvalidProgramException` |
| `TryGetValue("a", out got)` into an `int32?` slot | 2 `StackUnexpected`; `InvalidProgramException` |
| `for k, v in map[string, int32?]` | 1 `StackUnexpected`; `InvalidProgramException` |
| `for v in m.Values` over `map[string, int32?]` | 1 `StackUnexpected` |
| `m["a"] = nil` then `m["a"].HasValue` | 2 `StackUnexpected`; `InvalidProgramException` |
| `m.ContainsValue(9)` over `map[string, int32?]` | 1 `StackUnexpected` |
| `map[string, DateTime?]` write, read, `.Value.Year` | 2 `StackUnexpected` |
| `map[string, ImportedPoint?]` write and `.Value.X` | 2 `StackUnexpected` |
| `map[int32?, string]` key position | 2 `StackUnexpected`; `InvalidProgramException` |
| `map[DateTime?, string]` key position | 1 `StackUnexpected` |
| `var d Dictionary[string, int32?] = map[string, int32?]{…}` | 2 `StackUnexpected`; `InvalidProgramException` |
| an open `map[K, V]` parameter called with a `map[string, int32?]` | red — the closed *argument*'s backing dictionary reached the call |
| `for v in m.Values` over `map[string, DateTime?]` | red |
| `for k in m.Keys` over `map[DateTime?, string]` | red |
| `for k, v in m` over `map[string, DateTime?]` | red |
| the 5 control rows | green |

Reverting `MapTypeSymbol.cs`, `MemberLookup.cs` and `MethodBodyEmitter.MemberAccess.cs` to `3e5103de`, rebuilding, and running `--filter FullyQualifiedName~Issue4035`:

```
Failed! - Failed: 16, Passed: 5, Skipped: 0, Total: 21
```

**16 red / 5 green before, all 21 green after.** Every accepted row turns red on the revert and every control row stays green, so the fix is what moved them.

Nullable **reference** values (`map[string, string?]`) were never affected — the annotation has no distinct runtime type there, so `Dictionary<string, string>` is the correct backing — and are a control row that an over-broad fix would break.

---

## What located the layer

For the *same* `Dictionary[string, string]`, the indexer, `for k, v in m` and `TryGetValue` all agreed on `string` while only `.Keys`/`.Values` said `string?`. Three paths agreeing against one is what said the defect was in property-type derivation rather than in the #1354 flag rule, and it is the single most useful fact for whoever reads this next.

---

### CI: `tests (interpreter)` — the failure was mine, and the fix is a fixture that had encoded the bug

Discriminating measurement: `Adr0158SyncMapSpikeTests` is **6/6 green on `3e5103de`** and **4/6 red on this branch**, deterministically (three runs each, not flaky).

The diagnostic the log truncates is:

```
GS0536: Redundant '!!': the value is already non-null here.
```

a **warning**, which the suite's `AssertNoRealDiagnostics` refuses (it allows only `GS0286`).

Its source is the shared `SyncMapPrototype` fixture:

```gs
func Keys() []string {
    lock items {
        var ks = System.Collections.Generic.List[string]()
        for k in items.Keys {
            ks.Add(k!!)          // <-- here
        }
        return ks.ToArray()
    }
}
```

That `!!` was **load-bearing only because of the bug this PR fixes**: `items.Keys` handed the loop a spurious `string?`. With the element correctly `string` the assertion is genuinely redundant, and the compiler correctly says so. This is the fix working, surfacing in a fixture that had been written around the defect.

I dropped the `!!` rather than suppressing the warning — dropping it is what *proves* the defect is gone. Measured: the de-asserted fixture is **green on `3e5103de` too**, so the edit masks nothing on either side. It sits in commit 1 (#4033), the commit that makes `.Keys` non-nullable, with a `<para>` on the fixture recording why.

Blast radius of that diagnostic, measured rather than grepped: the **entire** `Interpreter.Tests` assembly was run, and exactly those 4 tests failed — all four share the one fixture; 1512 others passed. After the fix, **1516/1516**.

**Lesson taken:** `Interpreter.Tests` is now in the sweep. My original sweep covered `Compiler.Tests` and `Core.Tests` thoroughly and simply never ran the third assembly — precisely where a member-lookup change lands when the receiver is a runtime concurrent map.

### Finding 1 — untested non-host reflection-context branch → **fixed, and mutation-verified**

Correct, and the gap was real: that arm needs a non-host erasure context, and only *iterating* an imported nullable reaches it — indexing one never calls `ProjectSymbolicArgToErasedClr`. Three IL-verified rows added (`.Values` over `map[string, DateTime?]`, `.Keys` over `map[DateTime?, string]`, and `for k, v in m`), each moving a value through `v.Value.Year`.

Rather than assume they cover it, I mutation-tested the arm: replacing its lookup with `null` leaves the `int32?` iteration row green (it takes the host arm) and turns all three new rows into `GS0158: Cannot find member Value`. That evidence is recorded in the code comment on the rows.

### Finding 2 — stale row count → **fixed at the cause**

The `11 red rows in all` was a leftover from the ad-hoc probe matrix that predated the suite, so it described a different row set than a reverter would see. Rather than bump it to 15 — which would drift again on the next added row — I removed the count from that paragraph entirely and left the discrimination-witness paragraph as the single place that states one; the paragraph now says so explicitly. Both new test files were swept: no hard-coded row counts remain outside the witness paragraphs.

### Finding 3 — "two … per program" overstates it → **fixed**

Visible in my own witness table: `.Values`, `ContainsValue`, `for k, v in m` and the nullable `DateTime?` key each show one `StackUnexpected`. Scoped to the headline repro in the release notes, and the same overstatement corrected in the test file's remarks.

---

## Verification

| suite | result |
| --- | --- |
| `Compiler.Tests` (Nullab, Issue1354/3705/794/209/2176/3522/3311/3415/4011/4012/4015/4016/4020/4023/4024/4026/4033/4035, Map, Dictionary, ForIn, Iteration, ContinueWith, Issue3093) | **1086 passed, 0 failed** |
| `Core.Tests` (Nullab, ClrNullability, Issue1354/3705/834/794/209/2176/3522, Map, Dictionary, MemberLookup, Erasure, Symbols) | **1317 passed, 0 failed** |
| `Interpreter.Tests` (**entire assembly**) | **1516 passed, 0 failed** |
| `Issue4033…Tests` + `Issue4035…Tests` alone | 40 passed (19 + 21) |
| cs2gs code exploder | `17/17 apps green; run PASSED.` |

Every accepted and control row **executes and IL-verifies** (`IlVerifier.Verify`), and every green row moves a value through the member so a wrong binding is observably wrong. Rejected rows assert diagnostic **counts**, not presence.

`Issue3764CrossContextValueTypeCallShapeTests` and `Issue3755PinningShapeTargetFrameworkTests` were not run — known-environmental, missing `NETStandard.Library.Ref` locally.

## Docs

Both newly-working shapes are in `website/docs/release-notes.md` under **Fixed**. No new diagnostic is minted, so there is no `docs/diagnostics.md` / `website/docs/ref/diagnostics.md` row to keep in sync. No ADR-0174 erratum: neither issue is channel business.

## Out of scope, measured — filed separately

- **A method group in value position reaches emit with no diagnostic (GS9998).** The "missing diagnostic underneath" #4033 mentions. `l.Count` on a `List[int32]?` binds LINQ's `Enumerable.Count` extension *group*, and `.ToString()` on a method group reaches emit unreported. It reproduces with no nullable in sight — `List[int32]().Add.ToString()` on a plain non-nullable receiver gives the identical GS9998 — so it is neither this bug nor a nullability bug, and choosing the honest message is a design call.
- **A receiver's own `[Nullable]` annotation is dropped by member projection.** `GetProjectionReceiverImportedType` unwraps a `NullabilityAnnotatedTypeSymbol` to its bare base, discarding the per-position flags that `GetClrFieldTypeSymbol` alone rebuilds. Measured on `3e5103de`, a `Dictionary<string, string?>` from a nullable-enabled assembly already lost its `string?` through the indexer and through `for k, v`; this PR makes `.Values` agree with them rather than changing it.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01UVZ9GDN9g4NVF47GRNpWkL

### CODE REVIEW

File: src/Core/CodeAnalysis/Binding/MemberLookup.cs [hunk 1]
```diff
@@ -5849,6 +5937,46 @@ private static Type ResolveErasedObjectInContext(Type openDef)
...
+        {
+            // The underlying erased to the `object` placeholder (or to a
+            // reference type); `Nullable<>` cannot close over it, and the bare
+            // erasure is the shape every other reader already expects.
+            return underlying;
+        }
+
+        var openNullable = contextObject.Assembly == typeof(object).Assembly
+            ? typeof(Nullable<>)
+            : contextObject.Assembly.GetType("System.Nullable`1", throwOnError: false);
```
[the remark this answers is not in the document]
Comment (DavidObando) [reply src/Core/CodeAnalysis/Binding/MemberLookup.cs:5964]: Agreed, and measured — the branch was genuinely uncovered.

The arm needs an erasure context whose assembly is *not* the host corelib, and only **iterating** an imported nullable gets there; indexing one never calls `ProjectSymbolicArgToErasedClr`. So the `DateTime?` rows I had could not reach it, and the `int32?` iteration row takes the host arm.

Three rows added, all executing and IL-verified:

- `iterate-the-values-of-a-map-over-a-nullable-imported-struct` — `for v in m.Values` over `map[string, DateTime?]`, reading `v.Value.Year`
- `iterate-the-keys-of-a-map-over-a-nullable-imported-struct-key` — the `.Keys` twin over `map[DateTime?, string]`
- `iterate-key-and-value-of-a-map-over-a-nullable-imported-struct` — the `for k, v in m` form

Rather than assume they cover it, I **mutation-tested** the arm: replacing its lookup with `null` leaves the `int32?` iteration row green (its context *is* the host, so it takes the other arm) and turns all three new rows into `GS0158: Cannot find member Value`. That result is recorded in the code comment on the rows, so the next reader does not have to redo it.

All three are also red on `3e5103de`, which moved the #4035 witness from 13 red / 5 green to **16 red / 5 green, all 21 green after**.

File: test/Compiler.Tests/Issue4035MapOverANullableValueTypeTests.cs [hunk 2]
```diff
@@ -0,0 +1,645 @@
...
+/// <c>var d Dictionary[string, int32?] = map[string, int32?]{…}</c> failed with
+/// <c>[found ref Dictionary`2&lt;string,int32&gt;][expected ref
+/// Dictionary`2&lt;string,Nullable`1&lt;int32&gt;&gt;]</c>. That row is in
+/// <see cref="AcceptedCases"/>; it is the one that proves the fix made the
+/// spellings agree rather than merely moving the map.</para>
+/// <para><b>Measured scope.</b> Every value-position operation was affected —
+/// literal, <c>m[k]</c> read, <c>m[k] = v</c>, <c>m[k] = nil</c>,
+/// <c>TryGetValue</c>, <c>ContainsValue</c>, <c>for k, v in m</c>,
+/// <c>for v in m.Values</c> — and so was the KEY position
+/// (<c>map[int32?, V]</c>, <c>map[DateTime?, V]</c>): 11 red rows in all.
```
[the remark this answers is not in the document]
Comment (DavidObando) [reply test/Compiler.Tests/Issue4035MapOverANullableValueTypeTests.cs]: Fixed, and fixed at the cause rather than the symptom.

That `11 red rows in all` was a leftover from the ad-hoc probe matrix I ran before the suite existed, so it was describing a different set of rows than the one a reverter would actually see — exactly the misleading-revert-check risk you name. Rather than bump it to the new number (which is now 15, and would drift again the moment a row is added), I **deleted the count from that paragraph** and left the discrimination-witness paragraph as the single place that states one. The paragraph now says so explicitly, so the next editor knows where the number lives.

I swept both new test files for the same pattern afterwards; no other hard-coded row counts remain outside the witness paragraphs.

The witness number itself is re-measured, not arithmetic: reverting the three source files to `3e5103de` and running `--filter FullyQualifiedName~Issue4035` gives `Failed: 16, Passed: 5, Total: 21`.

File: website/docs/release-notes.md [hunk 3]
```diff
@@ -70,6 +70,8 @@ G# is pre-1.0. The repository's version base is currently `0.4`, and product ver
...
...w `InvalidProgramException`, because its backing `Dictionary` was closed over the bare `int32` while the literal, the indexer's `out` slot and the loop variable all carried a `Nullable<int32>` — two ILVerify `StackUnexpected` errors per program. A nullable type symbol answers its underlying type's CLR type, since a nullable *reference* has no distinct runtime type, and the map's construction read that directly instead of the `GetEffectiveClrType` correction its `sequence[T]` sibling already used. Every value-position operation was affected — the literal, `m[k]`, `m[k] = v`, `m[k] = nil`, `TryGetValue`, `ContainsValue`, `for k, v in m` and `for v in m.Values` — and so was the key position, `map[int32?, V]` and `map[DateTime?, V]` alike. Over an imported underlying such as `DateTime` the wrapper is now built inside that type's own reflection context, alongside the open `Dictionary`, for the reason issue #4023 documents. The `Dictionary[string, int32?]` spelling was already correct, so the two spellings of one type disagreed and assigning one to the other failed to verify; they now agree. Nullable *reference* values such as `map[string, string?]` were never affected and are unchanged.
```
[the remark this answers is not in the document]
Comment (DavidObando) [reply website/docs/release-notes.md]: Correct — you can see it in my own witness table: `.Values`, `ContainsValue`, `for k, v in m` and the nullable `DateTime?` key each show one `StackUnexpected`, not two. The count tracks how many nullable values the program moves, not the shape of the bug.

Scoped to the headline repro rather than generalised:

> … all carried a `Nullable<int32>` — two ILVerify `StackUnexpected` errors in that repro, one or more in every affected shape.

I applied the same correction to the matching sentence in the test file's remarks, which had the same overstatement (`Every row compiles … and then fails ilverify with `StackUnexpected``, now "with one or more `StackUnexpected` errors … the count per program varies with how many nullable values the row moves — the headline indexer repro shows two, a bare `.Values` iteration one").
